diff -ruN linux-2.4.19.orig/arch/s390/kernel/head.S linux-2.4.19/arch/s390/kernel/head.S
--- linux-2.4.19.orig/arch/s390/kernel/head.S	Mon Feb 17 05:59:50 2003
+++ linux-2.4.19/arch/s390/kernel/head.S	Tue Mar  4 23:59:18 2003
@@ -256,6 +256,7 @@
 #endif  /* CONFIG_IPL_VM */
 
 iplstart:
+        stm   %r0,%r15,0x180                   # store all registers
         lh    %r1,0xb8                         # test if subchannel number
         bct   %r1,.Lnoload                     #  is valid
 	l     %r1,0xb8                         # load ipl subchannel number
@@ -459,7 +460,16 @@
 # or linload or SALIPL
 #
         .org  0x10000
-startup:basr  %r13,0                     # get base
+startup:
+#ifdef CONFIG_SHARED_KERNEL
+#
+# Allows use of the IPL PARM parameter with NSSes and remains compatable
+# with none NSS usage.
+#
+        lm    %r0,%r15,0x180             # load ipl regs
+        stm   %r0,%r15,0x180             # (re)save regs
+#endif
+        basr  %r13,0                     # get base
 .LPG1:  lctl  %c0,%c15,.Lctl-.LPG1(%r13) # load control registers
 	la    %r12,_pstart-.LPG1(%r13)   # pointer to parameter area
 					 # move IPL device to lowcore
diff -ruN linux-2.4.19.orig/arch/s390/kernel/setup.c linux-2.4.19/arch/s390/kernel/setup.c
--- linux-2.4.19.orig/arch/s390/kernel/setup.c	Mon Feb 17 05:59:19 2003
+++ linux-2.4.19/arch/s390/kernel/setup.c	Wed Mar  5 00:03:58 2003
@@ -35,11 +35,13 @@
 #include <linux/bootmem.h>
 #include <linux/console.h>
 #include <linux/seq_file.h>
+#include <linux/ctype.h>
 #include <asm/uaccess.h>
 #include <asm/system.h>
 #include <asm/smp.h>
 #include <asm/mmu_context.h>
 #include <asm/cpcmd.h>
+#include <asm/ebcdic.h>
 
 /*
  * Machine setup..
@@ -69,6 +71,13 @@
 
 #include <asm/setup.h>
 
+/*
+ * Change this to whatever you want the VM PARM eyecatcher to be or set to
+ * the null ("") string to disable.  (It must be in uppercase EBCDIC.)
+ */
+#define PARMEYEC "\xd3\xc9\xd5\xc5\x40"
+#define PARMEYEL ((int)sizeof(PARMEYEC)-1)
+
 static char command_line[COMMAND_LINE_SIZE] = { 0, };
        char saved_command_line[COMMAND_LINE_SIZE];
 
@@ -340,6 +349,28 @@
 	code_resource.end = (unsigned long) &_etext - 1;
 	data_resource.start = (unsigned long) &_etext;
 	data_resource.end = (unsigned long) &_edata - 1;
+
+        /* Append VM PARMs */
+        if (MACHINE_IS_VM && (memcmp(IPL_REGS, PARMEYEC, PARMEYEL) == 0)) {
+                int last = 0;
+
+		/* Find last nonblank character */
+                for (i = 0; i < COMMAND_LINE_SIZE; i++) {
+                        if (from[i] == '\0')
+                                break;
+                        if (!isspace(from[i]))
+                                last = i + 1;
+                }
+
+                if (last < COMMAND_LINE_SIZE) {
+                        from[last++] = ' ';
+                        i = min(COMMAND_LINE_SIZE-last, 64-PARMEYEL);
+                        memcpy(&from[last], &((char *)IPL_REGS)[PARMEYEL], i);
+                        EBC_TOLOWER(&from[last], i);
+                        EBCASC(&from[last], i);
+                        from[last+i] = '\0';
+		}
+        }
 
         /* Save unparsed command line copy for /proc/cmdline */
         memcpy(saved_command_line, COMMAND_LINE, COMMAND_LINE_SIZE);
diff -ruN linux-2.4.19.orig/arch/s390x/kernel/head.S linux-2.4.19/arch/s390x/kernel/head.S
--- linux-2.4.19.orig/arch/s390x/kernel/head.S	Mon Feb 17 05:59:50 2003
+++ linux-2.4.19/arch/s390x/kernel/head.S	Tue Mar  4 22:38:59 2003
@@ -256,6 +256,7 @@
 #endif  /* CONFIG_IPL_VM */
 
 iplstart:
+        stm   %r0,%r15,0x180                   # store all registers
         lh    %r1,0xb8                         # test if subchannel number
         bct   %r1,.Lnoload                     #  is valid
 	l     %r1,0xb8                         # load ipl subchannel number
@@ -456,7 +457,16 @@
 # or linload or SALIPL
 #
         .org  0x10000
-startup:basr  %r13,0                     # get base
+startup:
+#ifdef CONFIG_SHARED_KERNEL
+#
+# Allows use of the IPL PARM parameter with NSSes and remains compatable
+# with none NSS usage.
+#
+        lm    %r0,%r15,0x180             # load ipl regs
+        stm   %r0,%r15,0x180             # (re)save regs
+#endif
+        basr  %r13,0                     # get base
 .LPG1:  sll   %r13,1                     # remove high order bit
         srl   %r13,1
         lhi   %r1,1                      # mode 1 = esame
diff -ruN linux-2.4.19.orig/arch/s390x/kernel/setup.c linux-2.4.19/arch/s390x/kernel/setup.c
--- linux-2.4.19.orig/arch/s390x/kernel/setup.c	Mon Feb 17 05:59:19 2003
+++ linux-2.4.19/arch/s390x/kernel/setup.c	Tue Mar  4 23:57:11 2003
@@ -35,11 +35,13 @@
 #include <linux/bootmem.h>
 #include <linux/console.h>
 #include <linux/seq_file.h>
+#include <linux/ctype.h>
 #include <asm/uaccess.h>
 #include <asm/system.h>
 #include <asm/smp.h>
 #include <asm/mmu_context.h>
 #include <asm/cpcmd.h>
+#include <asm/ebcdic.h>
 
 /*
  * Machine setup..
@@ -69,6 +71,13 @@
 
 #include <asm/setup.h>
 
+/*
+ * Change this to whatever you want the VM PARM eyecatcher to be or set to
+ * the null ("") string to disable.  (It must be in uppercase EBCDIC.)
+ */
+#define PARMEYEC "\xd3\xc9\xd5\xc5\x40"
+#define PARMEYEL ((int)sizeof(PARMEYEC)-1)
+
 static char command_line[COMMAND_LINE_SIZE] = { 0, };
        char saved_command_line[COMMAND_LINE_SIZE];
 
@@ -330,6 +339,28 @@
 	code_resource.end = (unsigned long) &_etext - 1;
 	data_resource.start = (unsigned long) &_etext;
 	data_resource.end = (unsigned long) &_edata - 1;
+
+        /* Append VM PARMs */
+        if (MACHINE_IS_VM && (memcmp(IPL_REGS, PARMEYEC, PARMEYEL) == 0)) {
+                int last = 0;
+
+		/* Find last nonblank character */
+                for (i = 0; i < COMMAND_LINE_SIZE; i++) {
+                        if (from[i] == '\0')
+                                break;
+                        if (!isspace(from[i]))
+                                last = i + 1;
+                }
+
+                if (last < COMMAND_LINE_SIZE) {
+                        from[last++] = ' ';
+                        i = min(COMMAND_LINE_SIZE-last, 64-PARMEYEL);
+                        memcpy(&from[last], &((char *)IPL_REGS)[PARMEYEL], i);
+                        EBC_TOLOWER(&from[last], i);
+                        EBCASC(&from[last], i);
+                        from[last+i] = '\0';
+		}
+        }
 
         /* Save unparsed command line copy for /proc/cmdline */
         memcpy(saved_command_line, COMMAND_LINE, COMMAND_LINE_SIZE);
diff -ruN linux-2.4.19.orig/include/asm-s390/setup.h linux-2.4.19/include/asm-s390/setup.h
--- linux-2.4.19.orig/include/asm-s390/setup.h	Mon Feb 25 13:38:13 2002
+++ linux-2.4.19/include/asm-s390/setup.h	Sun Mar  2 14:01:11 2003
@@ -19,6 +19,7 @@
 #define INITRD_START      (*(unsigned long *)  (0x1040C))
 #define INITRD_SIZE       (*(unsigned long *)  (0x10414))
 #define COMMAND_LINE      ((char *)            (0x10480))
+#define IPL_REGS          ((unsigned int *)    (0x180))
 
 /*
  * Machine features detected in head.S
@@ -53,6 +54,7 @@
 #define INITRD_START      0x1040C
 #define INITRD_SIZE       0x10414
 #define COMMAND_LINE      0x10480
+#define IPL_REGS          0x180
 
 #endif
 
diff -ruN linux-2.4.19.orig/include/asm-s390x/setup.h linux-2.4.19/include/asm-s390x/setup.h
--- linux-2.4.19.orig/include/asm-s390x/setup.h	Mon Feb 17 05:59:21 2003
+++ linux-2.4.19/include/asm-s390x/setup.h	Sun Mar  2 14:01:48 2003
@@ -19,6 +19,7 @@
 #define INITRD_START      (*(unsigned long *)  (0x10408))
 #define INITRD_SIZE       (*(unsigned long *)  (0x10410))
 #define COMMAND_LINE      ((char *)            (0x10480))
+#define IPL_REGS          ((unsigned int *)    (0x180))
 
 /*
  * Machine features detected in head.S
@@ -52,6 +53,7 @@
 #define INITRD_START      0x10408
 #define INITRD_SIZE       0x10410
 #define COMMAND_LINE      0x10480
+#define IPL_REGS          0x180
 
 #endif
 
